### PR TITLE
Added UserStatus check to gateway

### PR DESCRIPTION
**Pull request by sanyavertolet**
This PR is a part of #2451 

### What's done:
 * Added `status` field to `SaveUserDetails`
 * Forbade most endpoints to be used by inactive users

### CODE REVIEW

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt [hunk 1]
```diff
@@ -69,20 +77,12 @@ class WebSecurityConfig(
...
-                        } else {
-                            it
-                        }
-                    }
+                .pathMatchers(*allowedForInactiveEndpoints.toTypedArray()).access(::defaultAuthorizationDecision)
+                .pathMatchers("/api/**").access { authorization, authorizationContext ->
+                    authorization.flatMap { backendService.findByName(it.name) }
+                        .filter { it.isActive() }
+                        .flatMap { defaultAuthorizationDecision(authorization, authorizationContext) }
+                        .switchIfEmpty { AuthorizationDecision(false).toMono() }
```
Comment (nulls) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:85]:
```suggestion
                        .defaultIfEmpty(AuthorizationDecision(false))
```

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt [hunk 2]
```diff
@@ -15,18 +16,21 @@ import org.springframework.security.web.authentication.preauth.PreAuthenticatedA
...
  * @property name [com.saveourtool.save.entities.User.name]
  * @property role [com.saveourtool.save.entities.User.role]
+ * @property status [com.saveourtool.save.entities.User.status]
  * @property token [com.saveourtool.save.entities.User.password]
  */
 class SaveUserDetails(
     val id: Long,
     val name: String,
     val role: String,
+    val status: String,
```
Comment (nulls) [inline authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt:26]: why String and not `UserStatus`?

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt (hunk 1 above)
Comment (sanyavertolet) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt]: Done

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt [hunk 3]
```diff
@@ -45,8 +49,14 @@ class SaveUserDetails(
         httpHeaders.set(AUTHORIZATION_ID, id.toString())
         httpHeaders.set(AUTHORIZATION_NAME, name)
         httpHeaders.set(AUTHORIZATION_ROLES, role)
+        httpHeaders.set(AUTHORIZATION_STATUS, status)
     }
 
+    /**
+     * @return true if [status] is [UserStatus.ACTIVE], false otherwise
+     */
+    fun isActive() = status == UserStatus.ACTIVE.toString()
```
Comment (nulls) [inline authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt:58]: suggestion to change implementation for `isEnabled` to check status

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SecurityUtils.kt [hunk 4]
```diff
@@ -42,3 +64,26 @@ fun roleHierarchy(): RoleHierarchy = mapOf(
...
     }
+
+/**
+ * Get default [AuthorizationDecision] by [authentication] and [authorizationContext]
+ *
+ * @param authentication
+ * @param authorizationContext
+ * @return [Mono] of [AuthorizationDecision]
+ */
+fun defaultAuthorizationDecision(
```
Comment (nulls) [inline authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SecurityUtils.kt:75]: why do you extract this method?
maybe return back as a private function?

Comment (sanyavertolet) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SecurityUtils.kt:75]: private function is fine, will fix it

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt (hunk 3 above)
Comment (sanyavertolet) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt:58]: Will done for now, bit later it seems that we would support several user statuses but not only `ACTIVE`

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt (hunk 2 above)
Comment (sanyavertolet) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt:26]: Do not want to `find` required `UserStatus` when parsing headers

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt (hunk 3 above)
Comment (sanyavertolet) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt]: Done

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SecurityUtils.kt (hunk 4 above)
Comment (sanyavertolet) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SecurityUtils.kt]: Done

File: authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt (hunk 2 above)
Comment (nulls) [reply authentication-service/src/main/kotlin/com/saveourtool/save/authservice/utils/SaveUserDetails.kt:26]: yeap, you're right. forget about headers

**Review by nulls — APPROVED** [review]

**Pull request merged**